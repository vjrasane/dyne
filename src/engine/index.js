export { default as engine } from "~/engine/core";
export { Cmd } from "~/engine/effects/command";
export { Sub } from "~/engine/effects/subscription";
export { Update, Update as Init } from "~/engine/core/update";
