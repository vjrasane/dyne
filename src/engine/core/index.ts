import { renderer, VirtualDom } from "../render";
import { getInitializer, Initializer, Init } from "../core/init";
import { getUpdater, Updater, Update } from "../core/update";
import { getViewer, View, Viewer } from "../core/view";
import { dispatch, queues } from "../core/dispatch";
import { Sink } from "../../utils";

const UPDATE_INTERVAL = 1; /* milliseconds */
const VIEW_INTERVAL = 10; /* milliseconds */
const RENDER_INTERVAL = 10; /* milliseconds */
const EFFECT_INTERVAL = 1; /* milliseconds */

const loop = (procedure: Sink, interval: number): void => {
  let lock = false;
  setInterval((): void => {
    if (!lock) {
      lock = true;
      procedure();
      lock = false;
    }
  }, interval);
};

type Setup<M> = {
  init: Init<M>;
  update: Update<M>;
  view: View<M>;
};

const engine = <M>({ init, update, view }: Setup<M>) => {
  const updater: Updater<M> = getUpdater(update);
  const viewer: Viewer<M> = getViewer(view);
  const initializer: Initializer<M> = getInitializer(init);

  return ({ node, flags }) => {
    let model: M = initializer(flags);

    // update loop
    loop(() => {
      if (queues.msg.length > 0) {
        model = updater(queues.msg.shift(), model);
      }
    }, UPDATE_INTERVAL);

    // view loop
    let initialRender: Boolean = false;
    let viewedModel: M;
    let virtualDom: VirtualDom;
    const render = () => {
      virtualDom = viewer(model);
      viewedModel = model;
    };

    loop(() => {
      if (!initialRender) {
        render();
        initialRender = true;
      } else if (model && model !== viewedModel) {
        render();
      }
    }, VIEW_INTERVAL);

    // render loop
    let renderedVirtualDom;
    loop(() => {
      if (virtualDom && virtualDom !== renderedVirtualDom) {
        renderer(node, virtualDom, renderedVirtualDom);
        renderedVirtualDom = virtualDom;
      }
    }, RENDER_INTERVAL);

    // effect loop
    loop(() => {
      if (queues.effect.length > 0) {
        const effect = queues.effect.shift();
        // execute effect with the message dispatcher
        effect.execute(dispatch);
      }
    }, EFFECT_INTERVAL);
  };
};

export default engine;
