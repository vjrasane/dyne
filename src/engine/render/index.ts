import { isFunction, exists } from "../utils";
import { msgEventListener } from "../effects/command";
import { isElement, DyneElement } from "../../element";
import { dispatch } from "../core/dispatch";

const isEventProp = (name: string): boolean => /^on/.test(name);

const extractEventName = (name: string): string => name.slice(2).toLowerCase();

const removeBooleanProp = ($target: HTMLElement, name: string): void => {
  $target.removeAttribute(name);
  $target[name] = false;
};

const removeProp = (
  $target: HTMLElement,
  name: string,
  value: string | boolean
) => {
  if (name === "className") {
    $target.removeAttribute("class");
  } else if (typeof value === "boolean") {
    removeBooleanProp($target, name);
  } else {
    $target.removeAttribute(name);
  }
};

const setBooleanProp = ($target: HTMLElement, name: string, value) => {
  if (value) {
    $target.setAttribute(name, value);
    $target[name] = true;
  } else {
    $target[name] = false;
  }
};

const propChanged = (firstProp: Prop, secondProp: Prop) =>
  !(
    firstProp === secondProp ||
    (isFunction(firstProp) &&
      isFunction(secondProp) &&
      firstProp.toString() === secondProp.toString())
  );

const nodeChanged = (firstNode: DomElement, secondNode: VirtualDom) =>
  !(
    firstNode === secondNode ||
    (firstNode instanceof DyneElement &&
      secondNode instanceof DyneElement &&
      firstNode.type === secondNode.type)
  );

type Prop = string | boolean | Function;

const setProp = ($target: HTMLElement, name: string, value): void => {
  if (isEventProp(name)) {
    $target.addEventListener(
      extractEventName(name),
      msgEventListener(value, dispatch)
    );
  } else if (name === "className") {
    $target.setAttribute("class", value);
  } else if (typeof value === "boolean") {
    setBooleanProp($target, name, value);
  } else {
    $target.setAttribute(name, value);
  }
};

const setProps = ($target: HTMLElement, props: object): void => {
  Object.keys(props).forEach(name => {
    setProp($target, name, props[name]);
  });
};

const updateProp = (
  $target: HTMLElement,
  name: string,
  newValue: string,
  oldValue: string
) => {
  if (!newValue) {
    removeProp($target, name, oldValue);
  } else if (!oldValue || propChanged(newValue, oldValue)) {
    setProp($target, name, newValue);
  }
};

const updateProps = (
  $target: HTMLElement,
  newProps: object,
  oldProps: object = {}
) => {
  const props = Object.assign({}, newProps, oldProps);
  Object.keys(props).forEach(name => {
    updateProp($target, name, newProps[name], oldProps[name]);
  });
};

const createElement = (node: DyneElement | string) => {
  if (node instanceof DyneElement) {
    const $el = document.createElement(node.type);
    setProps($el, node.props);
    node.children.map(createElement).forEach($el.appendChild.bind($el));
    return $el;
  }
  return document.createTextNode(node);
};

type DomElement = string | DyneElement;

export type VirtualDom = DomElement | DomElement[];

const updateElement = (
  $parent: HTMLElement,
  newNode: DomElement,
  oldNode: VirtualDom,
  index = 0
) => {
  if (!exists(oldNode)) {
    $parent.appendChild(createElement(newNode));
  } else if (!exists(newNode)) {
    $parent.removeChild($parent.childNodes[index]);
  } else if (nodeChanged(newNode, oldNode)) {
    $parent.replaceChild(createElement(newNode), $parent.childNodes[index]);
  } else if (newNode instanceof DyneElement) {
    const oldElem =
      oldNode instanceof DyneElement ? oldNode : DyneElement.empty;
    updateProps(
      <HTMLElement>$parent.childNodes[index],
      newNode.props,
      oldElem.props
    );
    const newLength = newNode.children.length;
    const oldLength = oldElem.children.length;
    for (let i = 0; i < Math.max(newLength, oldLength); i++) {
      updateElement(
        <HTMLElement>$parent.childNodes[index],
        newNode.children[i],
        oldElem.children[i],
        i
      );
    }
  }
};

export const renderer = (
  $root: HTMLElement,
  newDom: VirtualDom,
  oldDom: VirtualDom
): void => {
  if (Array.isArray(newDom)) {
    newDom.forEach((node: DyneElement, index: number) =>
      updateElement($root, node, oldDom ? oldDom[index] : null, index)
    );
  } else {
    updateElement($root, newDom, oldDom);
  }
};
