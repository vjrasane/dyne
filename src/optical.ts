import { Update } from "./engine/core/update";
import { identity } from "./utils";

type Getter<P, C> = (parent: P) => C;

type Setter<P, C> = (child: C, parent: P) => P;

export class Lens<P, C> {
  get: Getter<P, C>;
  set: Setter<P, C>;
  constructor(get: Getter<P, C>, set: Setter<P, C>) {
    this.get = get;
    this.set = set;
  }

  static field = <P, C>(name: string): Lens<P, C> =>
    new Lens(getter(name), setter(name));

  static path = <P>(...path: string[]): Lens<P, any> =>
    path.reduce(
      <C>(lens: Lens<P, C>, field: string): Lens<P, any> =>
        lens.map(Lens.field(field)),
      Lens.identity
    );

  static identity: Lens<any, any> = new Lens(
    identity,
    (obj: any, _): any => obj
  );

  map = <S>(lens: Lens<C, S>): Lens<P, S> => {
    const get: Getter<P, S> = (parent: P) => lens.get(this.get(parent));
    const set: Setter<P, S> = (sub: S, parent: P) =>
      this.set(lens.set(sub, this.get(parent)), parent);
    return new Lens(get, set);
  };
}

export type OpticalUpdate<P, C> = {
  lens: Lens<P, C>;
  update: Update<C>;
};

const getter = <P, C>(field: string) => (parent: P): C =>
  parent ? parent[field] : undefined;

const setter = <P, C>(field: string) => (child: C, parent: P): P => ({
  ...parent,
  [field]: child
});

export const Optical = <P, C>(
  lens: Lens<P, C>,
  update: Update<C>
): OpticalUpdate<P, C> => ({
  lens,
  update
});
