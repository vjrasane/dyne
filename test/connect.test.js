import { Cmd, Update } from "~/engine";
import { Lens } from "~/optical";

import connector from "~/connect";

const localUpdate = id => (msg, model) => ({
  ...model,
  id,
  msg,
  status: "updated"
});

describe("connect", () => {
  let connect, update;
  beforeEach(() => {
    ({ connect, update } = connector());
  });

  it("connects single update with field name", () => {
    connect(
      "field",
      localUpdate("single update")
    );

    const model = {
      field: {
        data: "single update"
      }
    };

    const updated = update("message", model);
    expect(updated).toMatchSnapshot();
  });

  it("connects single update with field path array", () => {
    connect(
      ["field"],
      localUpdate("single update")
    );

    const model = {
      field: {
        data: "single update"
      }
    };

    const updated = update("message", model);
    expect(updated).toMatchSnapshot();
  });

  it("connects single update with lens", () => {
    connect(
      Lens.field("field"),
      localUpdate("single update")
    );

    const model = {
      field: {
        data: "single update"
      }
    };

    const updated = update("message", model);
    expect(updated).toMatchSnapshot();
  });

  it("connects single update to missing field", () => {
    connect(
      "field",
      localUpdate("single update")
    );

    const updated = update("message", {});
    expect(updated).toMatchSnapshot();
  });

  it("connects single embedded update with field path array", () => {
    connect(
      ["firstChild", "secondChild", "thirdChild"],
      localUpdate("embedded child")
    );

    const model = {
      firstChild: {
        secondChild: {
          thirdChild: {
            data: "embedded child"
          }
        }
      }
    };

    const updated = update("message", model);
    expect(updated).toMatchSnapshot();
  });

  it("connects single embedded update with lens", () => {
    connect(
      Lens.path("firstChild", "secondChild", "thirdChild"),
      localUpdate("embedded child")
    );

    const model = {
      firstChild: {
        secondChild: {
          thirdChild: {
            data: "embedded child"
          }
        }
      }
    };

    const updated = update("message", model);
    expect(updated).toMatchSnapshot();
  });

  it("connects single embedded update to missing field", () => {
    connect(
      ["firstChild", "secondChild", "thirdChild"],
      localUpdate("single update")
    );

    const updated = update("message", {});
    expect(updated).toMatchSnapshot();
  });

  it("connects multiple updates", () => {
    connect(
      "firstChild",
      localUpdate("first child")
    );
    connect(
      ["secondChild"],
      localUpdate("second child")
    );
    connect(
      Lens.field("thirdChild"),
      localUpdate("third child")
    );

    const model = {
      firstChild: {
        data: "first child"
      },
      secondChild: {
        data: "second child"
      },
      thirdChild: {
        data: "third child"
      }
    };

    const updated = update("message", model);
    expect(updated).toMatchSnapshot();
  });

  it("connects complex updates", () => {
    connect(
      "firstChild",
      localUpdate("first child")
    );
    connect(
      ["secondChild", "thirdChild"],
      localUpdate("third child")
    );
    connect(
      Lens.field("fourthChild", "fifthChild", "sixthChild"),
      localUpdate("sixth child")
    );

    const model = {
      firstChild: {
        data: "first child"
      },
      secondChild: {
        thirdChild: {
          data: "third child"
        }
      },
      fourthChild: {
        fifthChild: {
          sixthChild: {
            data: "sixth child"
          }
        }
      }
    };

    const updated = update("message", model);
    expect(updated).toMatchSnapshot();
  });

  it("connects complex updates to missing fields", () => {
    connect(
      "firstChild",
      localUpdate("first child")
    );
    connect(
      ["secondChild", "thirdChild"],
      localUpdate("third child")
    );
    connect(
      Lens.path("fourthChild", "fifthChild", "sixthChild"),
      localUpdate("sixth child")
    );

    const updated = update("message", {});
    expect(updated).toMatchSnapshot();
  });
});
