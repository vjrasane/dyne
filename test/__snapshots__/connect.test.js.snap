// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connect connects complex updates 1`] = `
Return {
  "bind": [Function],
  "effects": Array [],
  "map": [Function],
  "model": Object {
    "firstChild": Object {
      "data": "first child",
      "id": "first child",
      "msg": "message",
      "status": "updated",
    },
    "fourthChild": Object {
      "fifthChild": Object {
        "sixthChild": Object {
          "data": "sixth child",
        },
      },
      "id": "sixth child",
      "msg": "message",
      "status": "updated",
    },
    "secondChild": Object {
      "thirdChild": Object {
        "data": "third child",
        "id": "third child",
        "msg": "message",
        "status": "updated",
      },
    },
  },
}
`;

exports[`connect connects complex updates to missing fields 1`] = `
Return {
  "bind": [Function],
  "effects": Array [],
  "map": [Function],
  "model": Object {
    "firstChild": Object {
      "id": "first child",
      "msg": "message",
      "status": "updated",
    },
    "fourthChild": Object {
      "fifthChild": Object {
        "sixthChild": Object {
          "id": "sixth child",
          "msg": "message",
          "status": "updated",
        },
      },
    },
    "secondChild": Object {
      "thirdChild": Object {
        "id": "third child",
        "msg": "message",
        "status": "updated",
      },
    },
  },
}
`;

exports[`connect connects multiple updates 1`] = `
Return {
  "bind": [Function],
  "effects": Array [],
  "map": [Function],
  "model": Object {
    "firstChild": Object {
      "data": "first child",
      "id": "first child",
      "msg": "message",
      "status": "updated",
    },
    "secondChild": Object {
      "data": "second child",
      "id": "second child",
      "msg": "message",
      "status": "updated",
    },
    "thirdChild": Object {
      "data": "third child",
      "id": "third child",
      "msg": "message",
      "status": "updated",
    },
  },
}
`;

exports[`connect connects single embedded update to missing field 1`] = `
Return {
  "bind": [Function],
  "effects": Array [],
  "map": [Function],
  "model": Object {
    "firstChild": Object {
      "secondChild": Object {
        "thirdChild": Object {
          "id": "single update",
          "msg": "message",
          "status": "updated",
        },
      },
    },
  },
}
`;

exports[`connect connects single embedded update with field path array 1`] = `
Return {
  "bind": [Function],
  "effects": Array [],
  "map": [Function],
  "model": Object {
    "firstChild": Object {
      "secondChild": Object {
        "thirdChild": Object {
          "data": "embedded child",
          "id": "embedded child",
          "msg": "message",
          "status": "updated",
        },
      },
    },
  },
}
`;

exports[`connect connects single embedded update with lens 1`] = `
Return {
  "bind": [Function],
  "effects": Array [],
  "map": [Function],
  "model": Object {
    "firstChild": Object {
      "secondChild": Object {
        "thirdChild": Object {
          "data": "embedded child",
          "id": "embedded child",
          "msg": "message",
          "status": "updated",
        },
      },
    },
  },
}
`;

exports[`connect connects single update to missing field 1`] = `
Return {
  "bind": [Function],
  "effects": Array [],
  "map": [Function],
  "model": Object {
    "field": Object {
      "id": "single update",
      "msg": "message",
      "status": "updated",
    },
  },
}
`;

exports[`connect connects single update with field name 1`] = `
Return {
  "bind": [Function],
  "effects": Array [],
  "map": [Function],
  "model": Object {
    "field": Object {
      "data": "single update",
      "id": "single update",
      "msg": "message",
      "status": "updated",
    },
  },
}
`;

exports[`connect connects single update with field path array 1`] = `
Return {
  "bind": [Function],
  "effects": Array [],
  "map": [Function],
  "model": Object {
    "field": Object {
      "data": "single update",
      "id": "single update",
      "msg": "message",
      "status": "updated",
    },
  },
}
`;

exports[`connect connects single update with lens 1`] = `
Return {
  "bind": [Function],
  "effects": Array [],
  "map": [Function],
  "model": Object {
    "field": Object {
      "data": "single update",
      "id": "single update",
      "msg": "message",
      "status": "updated",
    },
  },
}
`;
